federation/schema_syncer.go

```diff
@@ -27,43 +26,50 @@ func NewIntrospectionSchemaSyncer(ctx context.Context, executors map[string]Exec
 	return ss
 }
 
-func (s *IntrospectionSchemaSyncer) FetchPlanner(ctx context.Context) (*Planner, error) {
+func (s *IntrospectionSchemaSyncer) FetchPlannerAndIntrospectionQueryResult(ctx context.Context) (*Planner, []byte, error) {
 	schemas := make(map[string]*IntrospectionQueryResult)
+	versionedSchemas := make(map[string]map[string]*IntrospectionQueryResult)
 	for server, client := range s.executors {
 		resp, err := fetchSchema(ctx, client, s.queryMetadata)
 		if err != nil {
-			return nil, oops.Wrapf(err, "fetching schema %s", server)
+			return nil, nil, oops.Wrapf(err, "fetching schema %s", server)
 		}
 		schema := resp.Result
 		var iq IntrospectionQueryResult
 		if err := json.Unmarshal(schema, &iq); err != nil {
-			return nil, oops.Wrapf(err, "unmarshaling schema %s", server)
+			return nil, nil, oops.Wrapf(err, "unmarshaling schema %s", server)
 		}
 
 		schemas[server] = &iq
+		versionedSchemas[server] = map[string]*IntrospectionQueryResult{
+			"": &iq,
+		}
+	}
+
+	versionedSchemas, err := AddIntrospectionQueryToSchemaVersions(versionedSchemas)
+	if err != nil {
+		return nil, nil, oops.Wrapf(err, "adding introspection to schema")
 	}
 
-	types, err := convertSchema(schemas)
+	types, err := ConvertVersionedSchemas(versionedSchemas)
 	if err != nil {
-		return nil, oops.Wrapf(err, "converting schemas error")
+		return nil, nil, oops.Wrapf(err, "converting schemas error")
 	}
 
-	introspectionSchema := introspection.BareIntrospectionSchema(types.Schema)
-	schema, err := introspection.RunIntrospectionQuery(introspection.BareIntrospectionSchema(introspectionSchema))
-	if err != nil || schema == nil {
-		return nil, oops.Wrapf(err, "error running introspection query")
+	mergedIq, err := MergeIntrospectionSchemas(versionedSchemas)
+	if err != nil {
+		return nil, nil, oops.Wrapf(err, "merging introspection schemas")
 	}
 
-	var iq IntrospectionQueryResult
-	if err := json.Unmarshal(schema, &iq); err != nil {
-		return nil, oops.Wrapf(err, "unmarshaling introspection schema")
+	mergedIqBytes, err := json.Marshal(mergedIq)
+	if err != nil {
+		return nil, nil, oops.Wrapf(err, "marshalling introspection schema")
 	}
 
-	schemas["introspection"] = &iq
```

Comment:
It seems we can also just change this key to "introspectionclient" rather than merging the introspection of each sub schemas? Can you explain a bit more about why we need to query the introspection query separately?

Reply:
oh yeah you are right - will make that change